### PR TITLE
Guidance

### PR DESCRIPTION
**Pull request by MoKob**
refactor of the guidance system to allow easier integration of a new guidance system.

Whats left to do:

- [x] refactor existing description generation
- [x] get cucumber tests to work again
- [ ] allow for post processing of guidance

### CODE REVIEW

File: include/engine/guidance/api_response_generator.hpp [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+                                   static_cast<std::size_t>(result.size())};
+            result.emplace_back(std::move(seg));
+        }
+    }
+    return result;
+}
+
+template <typename DataFacadeT>
+osrm::json::Object
+ApiResponseGenerator<DataFacadeT>::buildHintData(const InternalRouteResult &raw_route) const
```
Comment (TheMarex) [inline include/engine/guidance/api_response_generator.hpp]: Methods start with capital letters, like `::BuildHintData`.

File: src/engine/guidance/segment_list.cc [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+
+#include <cstddef>
+
+#include <boost/assert.hpp>
+
+namespace osrm
+{
+namespace engine
+{
+namespace route
```
Comment (TheMarex) [inline src/engine/guidance/segment_list.cc]: the namespace should correspond with the filesystem structure. Like `engine/guidance/ ` -> `osrm::engine::guidance`.

**Comment by MoKob** [comment]
Whats left to do:

[x] refactor existing description generation
[ ] allow for post processing of guidance

File: include/engine/descriptors/descriptor_base.hpp [hunk 3]
```diff
@@ -1,6 +1,8 @@
 #ifndef DESCRIPTOR_BASE_HPP
 #define DESCRIPTOR_BASE_HPP
 
+#include "config.h"
```
Comment (TheMarex) [inline include/engine/descriptors/descriptor_base.hpp]: Don't use relative file paths. `engine/descriptors/config.hpp`

File: include/engine/guidance/api_response_generator.hpp [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+    : facade(facade_)
+{
+}
+
+template <typename DataFacadeT>
+void ApiResponseGenerator<DataFacadeT>::DescribeRoute(const DescriptorConfig &config,
+                                                      const InternalRouteResult &raw_route,
+                                                      osrm::json::Object &json_result)
+{
+	if( not raw_route.is_valid() ){
```
Comment (TheMarex) [inline include/engine/guidance/api_response_generator.hpp]: Letf over debug output?

File: include/engine/descriptors/config.h [hunk 5]
```diff
@@ -0,0 +1,21 @@
+#pragma once
```
Comment (daniel-j-h) [inline include/engine/descriptors/config.h]: I think `osrm-backend` consistently uses define-include guards instead of pragmas

File: include/engine/guidance/annotators/textual_route_annotation.hpp [hunk 6]
```diff
@@ -0,0 +1,133 @@
+#pragma once
```
Comment (daniel-j-h) [inline include/engine/guidance/annotators/textual_route_annotation.hpp]: same here and for the whole pr

File: include/engine/guidance/annotators/textual_route_annotation.hpp [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+template< typename DataFacadeT >
+inline osrm::json::Array
+AnnotateRoute(const std::vector<SegmentInformation> &route_segments, DataFacadeT* facade)
+{
+    osrm::json::Array json_instruction_array;
+    if( route_segments.empty() )
+      return json_instruction_array;
+    // Segment information has following format:
+    //["instruction id","streetname",length,position,time,"length","earth_direction",azimuth]
+    uint32_t necessary_segments_running_index = 0;
```
Comment (daniel-j-h) [inline include/engine/guidance/annotators/textual_route_annotation.hpp]: prefix with `std::`

File: include/engine/guidance/annotators/textual_route_annotation.hpp [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+
+    round_about = {std::numeric_limits<int32_t>::max(), 0, 0};
+    std::string temp_dist, temp_length, temp_duration, temp_bearing, temp_instruction;
+
+    //Generate annotations for every segment
+    for (const SegmentInformation &segment : route_segments)
+    {
+        osrm::json::Array json_instruction_row;
+        TurnInstruction current_instruction = segment.turn_instruction;
+        if (TurnInstructionsClass::TurnIsNecessary(current_instruction))
```
Comment (daniel-j-h) [inline include/engine/guidance/annotators/textual_route_annotation.hpp]: This is a rather long code block, can we split it? Or maybe even something along the lines of `for_each(route_segments | filtered(TurnNecessary), lambda)`

File: include/engine/guidance/annotators/textual_route_annotation.hpp [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+            }
+            else
+            {
+                std::string current_turn_instruction;
+                if (TurnInstruction::LeaveRoundAbout == current_instruction)
+                {
+                    temp_instruction =
+                        std::to_string(cast::enum_to_underlying(TurnInstruction::EnterRoundAbout));
+                    current_turn_instruction += temp_instruction;
+                    current_turn_instruction += "-";
```
Comment (daniel-j-h) [inline include/engine/guidance/annotators/textual_route_annotation.hpp]: Do we want to force this representation here in this function? Maybe make it a policy with a default value that makes sense.

File: include/engine/guidance/annotators/textual_route_annotation.hpp [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+                    current_turn_instruction += temp_instruction;
+                    round_about.leave_at_exit = 0;
+                }
+                else
+                {
+                    temp_instruction =
+                        std::to_string(cast::enum_to_underlying(current_instruction));
+                    current_turn_instruction += temp_instruction;
+                }
+                json_instruction_row.values.push_back(current_turn_instruction);
```
Comment (daniel-j-h) [inline include/engine/guidance/annotators/textual_route_annotation.hpp]: `std::move` it in, it is not needed from here on.

File: include/engine/guidance/annotators/textual_route_annotation.hpp (hunk 9 above)
Comment (MoKob) [inline include/engine/guidance/annotators/textual_route_annotation.hpp]: this function is pretty much just the old state. This is what we have to improve now that we have the rest reworked. Not really part of this PR.

File: include/engine/guidance/annotators/textual_route_annotation.hpp [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+
+                json_instruction_row.values.push_back(facade->get_name_for_id(segment.name_id));
+                json_instruction_row.values.push_back(std::round(segment.length));
+                json_instruction_row.values.push_back(necessary_segments_running_index);
+                json_instruction_row.values.push_back(std::round(segment.duration / 10.));
+                json_instruction_row.values.push_back(
+                    std::to_string(static_cast<uint32_t>(segment.length)) + "m");
```
Comment (daniel-j-h) [inline include/engine/guidance/annotators/textual_route_annotation.hpp]: Same here for the representation and all of the conversion: do we want to make the annotation generic in the sense that it takes policies e.g. a `InMeter` policy that does the conversion and representation? We could then alias common annotators, such as `MeterAnnotator` --- or is this overkill and we just hardcode it?

File: include/engine/guidance/annotators/textual_route_annotation.hpp [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+
+                // post turn bearing
+                const double post_turn_bearing_value = (segment.post_turn_bearing / 10.);
+                json_instruction_row.values.push_back(bearing::get(post_turn_bearing_value));
+                json_instruction_row.values.push_back(
+                    static_cast<uint32_t>(round(post_turn_bearing_value)));
```
Comment (daniel-j-h) [inline include/engine/guidance/annotators/textual_route_annotation.hpp]: Prefix `round` with `std::`!

File: include/engine/guidance/annotators/textual_route_annotation.hpp [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+
+                json_instruction_row.values.push_back(segment.travel_mode);
+
+                // pre turn bearing
+                const double pre_turn_bearing_value = (segment.pre_turn_bearing / 10.);
```
Comment (daniel-j-h) [inline include/engine/guidance/annotators/textual_route_annotation.hpp]: Would love to get meaningful conversion functions in here; also above for example `segment.duration / 10.` --- what about a small function / lambda `toSeconds` or the like?

File: include/engine/guidance/annotators/textual_route_annotation.hpp [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+        if (segment.necessary)
+        {
+            ++necessary_segments_running_index;
+        }
+    }
+
+    osrm::json::Array json_last_instruction_row;
+    temp_instruction =
+        std::to_string(cast::enum_to_underlying(TurnInstruction::ReachedYourDestination));
+    json_last_instruction_row.values.push_back(temp_instruction);
```
Comment (daniel-j-h) [inline include/engine/guidance/annotators/textual_route_annotation.hpp]: You can move `temp_instruction` in here.

File: include/engine/guidance/annotators/textual_route_annotation.hpp [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+    json_last_instruction_row.values.push_back("");
+    json_last_instruction_row.values.push_back(0);
+    json_last_instruction_row.values.push_back(necessary_segments_running_index - 1);
+    json_last_instruction_row.values.push_back(0);
+    json_last_instruction_row.values.push_back("0m");
```
Comment (daniel-j-h) [inline include/engine/guidance/annotators/textual_route_annotation.hpp]: Same hardcoded representation as outlined above ^

File: include/engine/guidance/annotators/textual_route_annotation.hpp [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+    json_last_instruction_row.values.push_back(bearing::get(0.0));
+    json_last_instruction_row.values.push_back(0.);
+    json_last_instruction_row.values.push_back(bearing::get(0.0));
+    json_last_instruction_row.values.push_back(0.);
+    json_instruction_array.values.push_back(json_last_instruction_row);
```
Comment (daniel-j-h) [inline include/engine/guidance/annotators/textual_route_annotation.hpp]: moveable

File: include/engine/guidance/api_response_generator.hpp [hunk 17]
```diff
@@ -0,0 +1,286 @@
...
+#include "util/typedefs.hpp"
+
+#include <boost/assert.hpp>
+
+#include <cstdint>
+#include <cstddef>
+#include <cmath>
+
+#include <limits>
+#include <vector>
```
Comment (daniel-j-h) [inline include/engine/guidance/api_response_generator.hpp]: string header needed

File: include/engine/guidance/annotators/textual_route_annotation.hpp [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+#include "extractor/turn_instructions.hpp"
+#include "osrm/json_container.hpp"
+#include "util/bearing.hpp"
+#include "util/cast.hpp"
+
+#include <cstdint>
+
+#include <limits>
+#include <string>
+#include <vector>
```
Comment (daniel-j-h) [inline include/engine/guidance/annotators/textual_route_annotation.hpp]: utility header for move

File: include/engine/guidance/api_response_generator.hpp [hunk 19]
```diff
@@ -0,0 +1,286 @@
...
+};
+} // namespace detail
+
+template <typename DataFacadeT> class ApiResponseGenerator
+{
+  public:
+    typedef DataFacadeT DataFacade;
+    typedef SegmentList<DataFacade> SegmentListT;
+    typedef detail::Segment Segment;
+    typedef ExtractRouteNames<DataFacade, Segment> RouteNameExtractor;
```
Comment (daniel-j-h) [inline include/engine/guidance/api_response_generator.hpp]: would just use `using` in the whole guidance subproject

File: include/engine/guidance/api_response_generator.hpp [hunk 20]
```diff
@@ -0,0 +1,286 @@
...
+    int32_t length;
+    std::size_t position;
+};
+} // namespace detail
+
+template <typename DataFacadeT> class ApiResponseGenerator
+{
+  public:
+    typedef DataFacadeT DataFacade;
+    typedef SegmentList<DataFacade> SegmentListT;
```
Comment (daniel-j-h) [inline include/engine/guidance/api_response_generator.hpp]: shouldn't it be `SegmentList` to be consistent with the other type aliases?

File: include/engine/guidance/api_response_generator.hpp [hunk 21]
```diff
@@ -0,0 +1,286 @@
...
+void ApiResponseGenerator<DataFacadeT>::DescribeRoute(const DescriptorConfig &config,
+                                                      const InternalRouteResult &raw_route,
+                                                      osrm::json::Object &json_result)
+{
+    if( not raw_route.is_valid() ){
+      return;
+    }
+    const bool ALLOW_SIMPLIFICATION = true;
+    const bool EXTRACT_ROUTE = false;
+    const bool EXTRACT_ALTERNATIVE = true;
```
Comment (daniel-j-h) [inline include/engine/guidance/api_response_generator.hpp]: can be static, or even constexpr (does this already work on Windows?)

File: include/engine/guidance/api_response_generator.hpp [hunk 22]
```diff
@@ -0,0 +1,286 @@
...
+
+    if (config.geometry)
+    {
+        json_result.values["route_geometry"] = GetGeometry(config.encode_geometry, segment_list);
+    }
+
+    if (config.instructions)
+    {
+        json_result.values["route_instructions"] =
+            annotators::AnnotateRoute(segment_list.Get(), facade);
```
Comment (daniel-j-h) [inline include/engine/guidance/api_response_generator.hpp]: This is the call site that would have to decide on the policy to use if we want to make the annotation generic

File: include/engine/guidance/api_response_generator.hpp [hunk 23]
```diff
@@ -0,0 +1,286 @@
...
+            SummarizeRoute(raw_route, alternate_segment_list));
+        json_result.values["alternative_summaries"] = json_alternate_route_summary_array;
+        json_result.values["alternative_indices"] = ListViaIndices(alternate_segment_list);
+
+        if (config.geometry)
+        {
+            osrm::json::Value alternate_geometry_string =
+                GetGeometry(config.encode_geometry, alternate_segment_list);
+            osrm::json::Array json_alternate_geometries_array;
+            json_alternate_geometries_array.values.push_back(alternate_geometry_string);
```
Comment (daniel-j-h) [inline include/engine/guidance/api_response_generator.hpp]: moveable

File: include/engine/guidance/api_response_generator.hpp [hunk 24]
```diff
@@ -0,0 +1,286 @@
...
+        osrm::json::Array json_alternate_route_summary_array;
+        json_alternate_route_summary_array.values.push_back(
+            SummarizeRoute(raw_route, alternate_segment_list));
+        json_result.values["alternative_summaries"] = json_alternate_route_summary_array;
+        json_result.values["alternative_indices"] = ListViaIndices(alternate_segment_list);
+
+        if (config.geometry)
+        {
+            osrm::json::Value alternate_geometry_string =
+                GetGeometry(config.encode_geometry, alternate_segment_list);
```
Comment (daniel-j-h) [inline include/engine/guidance/api_response_generator.hpp]: Hmm what I only now realize: I think you can `auto` away a lot of types here and throughout the codebase

File: include/engine/guidance/api_response_generator.hpp [hunk 25]
```diff
@@ -0,0 +1,286 @@
...
+        // generate names for both the main path and the alternative route
+        auto path_segments = BuildRouteSegments(segment_list);
+        auto alternate_segments = BuildRouteSegments(alternate_segment_list);
+        route_names = generate_route_names(path_segments, alternate_segments, facade);
+
+        osrm::json::Array json_alternate_names_array;
+        osrm::json::Array json_alternate_names;
+        json_alternate_names.values.push_back(route_names.alternative_path_name_1);
+        json_alternate_names.values.push_back(route_names.alternative_path_name_2);
+        json_alternate_names_array.values.push_back(json_alternate_names);
```
Comment (daniel-j-h) [inline include/engine/guidance/api_response_generator.hpp]: moveable

File: include/engine/guidance/api_response_generator.hpp [hunk 26]
```diff
@@ -0,0 +1,286 @@
...
+osrm::json::Object
+ApiResponseGenerator<DataFacadeT>::SummarizeRoute(const InternalRouteResult &raw_route,
+                                                  const SegmentListT &segment_list) const
+{
+    osrm::json::Object json_route_summary;
+    if (not raw_route.segment_end_coordinates.empty())
+    {
+        auto start_name_id = raw_route.segment_end_coordinates.front().source_phantom.name_id;
+        json_route_summary.values["start_point"] = facade->get_name_for_id(start_name_id);
+        auto destination_name_id = raw_route.segment_end_coordinates.back().target_phantom.name_id;
```
Comment (daniel-j-h) [inline include/engine/guidance/api_response_generator.hpp]: `const auto` ^

File: include/engine/guidance/api_response_generator.hpp [hunk 27]
```diff
@@ -0,0 +1,286 @@
...
+{
+    osrm::json::Array json_via_points_array;
+    osrm::json::Array json_first_coordinate;
+    json_first_coordinate.values.push_back(
+        raw_route.segment_end_coordinates.front().source_phantom.location.lat /
+        COORDINATE_PRECISION);
+    json_first_coordinate.values.push_back(
+        raw_route.segment_end_coordinates.front().source_phantom.location.lon /
+        COORDINATE_PRECISION);
+    json_via_points_array.values.push_back(json_first_coordinate);
```
Comment (daniel-j-h) [inline include/engine/guidance/api_response_generator.hpp]: moveable

File: include/engine/guidance/api_response_generator.hpp [hunk 28]
```diff
@@ -0,0 +1,286 @@
...
+    for (const PhantomNodes &nodes : raw_route.segment_end_coordinates)
+    {
+        std::string tmp;
+        osrm::json::Array json_coordinate;
+        json_coordinate.values.push_back(nodes.target_phantom.location.lat / COORDINATE_PRECISION);
+        json_coordinate.values.push_back(nodes.target_phantom.location.lon / COORDINATE_PRECISION);
+        json_via_points_array.values.push_back(json_coordinate);
```
Comment (daniel-j-h) [inline include/engine/guidance/api_response_generator.hpp]: moveable

File: include/engine/guidance/api_response_generator.hpp [hunk 29]
```diff
@@ -0,0 +1,286 @@
...
+}
+
+template <typename DataFacadeT>
+std::vector<detail::Segment>
+ApiResponseGenerator<DataFacadeT>::BuildRouteSegments(const SegmentListT &segment_list) const
+{
+    std::vector<detail::Segment> result;
+    for (const auto &segment : segment_list.Get())
+    {
+        auto current_turn = segment.turn_instruction;
```
Comment (daniel-j-h) [inline include/engine/guidance/api_response_generator.hpp]: can be const

File: include/engine/guidance/api_response_generator.hpp [hunk 30]
```diff
@@ -0,0 +1,286 @@
...
+    json_hint_object.values["checksum"] = facade->GetCheckSum();
+    osrm::json::Array json_location_hint_array;
+    std::string hint;
+    for (const auto i : osrm::irange<std::size_t>(0, raw_route.segment_end_coordinates.size()))
+    {
+        ObjectEncoder::EncodeToBase64(raw_route.segment_end_coordinates[i].source_phantom, hint);
+        json_location_hint_array.values.push_back(hint);
+    }
+    ObjectEncoder::EncodeToBase64(raw_route.segment_end_coordinates.back().target_phantom, hint);
+    json_location_hint_array.values.push_back(hint);
```
Comment (daniel-j-h) [inline include/engine/guidance/api_response_generator.hpp]: moveable

File: include/engine/guidance/segment_list.hpp [hunk 31]
```diff
@@ -0,0 +1,304 @@
...
+#include "osrm/coordinate.hpp"
+
+#include "engine/douglas_peucker.hpp"
+#include "engine/internal_route_result.hpp"
+#include "engine/phantom_node.hpp"
+#include "engine/segment_information.hpp"
+
+#include "extractor/turn_instructions.hpp"
+
+#include <vector>
```
Comment (daniel-j-h) [inline include/engine/guidance/segment_list.hpp]: cstdint
limits

File: include/engine/guidance/segment_list.hpp [hunk 32]
```diff
@@ -0,0 +1,304 @@
...
+                const bool extract_alternative,
+                const unsigned zoom_level,
+                const bool allow_simplification,
+                const DataFacade *facade);
+
+    const std::vector<std::uint32_t> &GetViaIndices() const;
+    std::uint32_t GetDistance() const;
+    std::uint32_t GetDuration() const;
+
+    std::vector<SegmentInformation> const &Get() const;
```
Comment (daniel-j-h) [inline include/engine/guidance/segment_list.hpp]: Can we be consistent about lhs vs rhs const placement?
Three lines above it is `const T &fn() const;` here it is `T const &fn() const;`

File: include/engine/guidance/segment_list.hpp [hunk 33]
```diff
@@ -0,0 +1,304 @@
...
+
+    void AppendSegment(const FixedPointCoordinate &coordinate, const PathData &path_point);
+    void Finalize(const bool extract_alternative,
+                  const InternalRouteResult &raw_route,
+                  const unsigned zoom_level,
+                  const bool allow_simplification);
+
+    // journey length in tenth of a second
+    std::uint32_t total_distance;
+    std::uint32_t total_duration;
```
Comment (daniel-j-h) [inline include/engine/guidance/segment_list.hpp]: For the duration... what about the distance? Dezimeters?

File: include/engine/guidance/segment_list.hpp [hunk 34]
```diff
@@ -0,0 +1,304 @@
...
+                                      const DataFacade *facade)
+    : total_distance(0), total_duration(0)
+{
+    if (not raw_route.is_valid()){
+        return;
+    }
+
+    if (extract_alternative)
+    {
+        BOOST_ASSERT( raw_route.has_alternative() );
```
Comment (daniel-j-h) [inline include/engine/guidance/segment_list.hpp]: clang-format should eliminate white spaces here

File: include/engine/guidance/segment_list.hpp [hunk 35]
```diff
@@ -0,0 +1,304 @@
...
+        if (path_point.segment_duration > 0)
+        {
+            segments.front().name_id = path_point.name_id;
+            segments.front().travel_mode = path_point.travel_mode;
+        }
+        return;
+    }
+
+    // make sure mode changes are announced, even when there otherwise is no turn
+    auto getTurn = [](const PathData &path_point, const TravelMode previous_mode)
```
Comment (daniel-j-h) [inline include/engine/guidance/segment_list.hpp]: const

File: include/engine/guidance/segment_list.hpp [hunk 36]
```diff
@@ -0,0 +1,304 @@
...
+        if (TurnInstruction::NoTurn == path_point.turn_instruction and
+            previous_mode != path_point.travel_mode and path_point.segment_duration > 0)
+        {
+            return TurnInstruction::GoStraight;
+        }
+        return path_point.turn_instruction;
+    };
+
+    //TODO check why we require .front() here
+    auto turn = segments.size() ? getTurn(path_point, segments.front().travel_mode) : path_point.turn_instruction;
```
Comment (daniel-j-h) [inline include/engine/guidance/segment_list.hpp]: const

File: include/engine/plugins/match.hpp [hunk 37]
```diff
@@ -185,55 +187,37 @@ template <class DataFacadeT> class MapMatchingPlugin : public BasePlugin
             subtrace.values["confidence"] = sub.confidence;
         }
 
-        JSONDescriptor<DataFacadeT> json_descriptor(facade);
-        json_descriptor.SetConfig(route_parameters);
+        osrm::engine::guidance::ApiResponseGenerator<DataFacadeT> response_generator(facade);
```
Comment (daniel-j-h) [inline include/engine/plugins/match.hpp]: small nitpicking, but we could create type inference `make*` helper functions, so that we can just say

    auto response_generator = makeApiResponseGenerator(facade);

File: include/engine/plugins/match.hpp [hunk 38]
```diff
@@ -185,55 +187,37 @@ template <class DataFacadeT> class MapMatchingPlugin : public BasePlugin
...
-            for (auto &segment : factory.path_description)
-            {
-                segment.necessary = true;
-            }
+            typedef osrm::engine::guidance::SegmentList<DataFacadeT> SegmentListT;
+            //Passing false to extract_alternative extracts the route.
+            const bool EXTRACT_ROUTE = false;
+            // by passing false to segment_list, we skip the douglas peucker simplification
+            // and mark all segments as necessary within the generation process
+            const bool NO_ROUTE_SIMPLIFICATION = false;
```
Comment (daniel-j-h) [inline include/engine/plugins/match.hpp]: static or even constexpr

File: include/engine/plugins/viaroute.hpp [hunk 39]
```diff
@@ -139,23 +138,23 @@ template <class DataFacadeT> class ViaRoutePlugin final : public BasePlugin
...
+            }
             break;
-        // case 2:
-        //      descriptor = osrm::make_unique<GEOJSONDescriptor<DataFacadeT>>();
-        //      break;
         default:
-            descriptor = osrm::make_unique<JSONDescriptor<DataFacadeT>>(facade);
+            {
+              osrm::engine::guidance::ApiResponseGenerator<DataFacadeT> generator( facade );
+              generator.DescribeRoute( route_parameters, raw_route, json_result );
```
Comment (daniel-j-h) [inline include/engine/plugins/viaroute.hpp]: clang-format

File: include/engine/guidance/segment_list.hpp [hunk 40]
```diff
@@ -0,0 +1,304 @@
...
+        {
+            BOOST_ASSERT(segments[i].necessary);
+            segment_length = 0;
+            segment_duration = 0;
+            segment_start_index = i;
+        }
+    }
+
+    total_distance = static_cast<std::uint32_t>(round(path_length));
+    total_duration = static_cast<std::uint32_t>(round(
```
Comment (daniel-j-h) [inline include/engine/guidance/segment_list.hpp]: Prefix with `std::`

File: include/engine/guidance/segment_list.hpp [hunk 41]
```diff
@@ -0,0 +1,304 @@
...
+    {
+        segments.erase(segments.begin());
+        segments.front().turn_instruction = TurnInstruction::HeadOn;
+        segments.front().necessary = true;
+        start_phantom.name_id = segments.front().name_id;
+    }
+
+    if (allow_simplification)
+    {
+        DouglasPeucker polyline_generalizer;
```
Comment (daniel-j-h) [inline include/engine/guidance/segment_list.hpp]: Can we cache and keep around a single `DouglasPeuker`? That is, making it static or storing outside.

File: include/engine/guidance/segment_list.hpp (hunk 32 above)
Comment (MoKob) [inline include/engine/guidance/segment_list.hpp]: yeah... have to force myself for the `const T &` version each time.

I prefer `T const& name` due to me preferring `T const* name` since `const` now is directly before the part that is `const`. Especially if you have `T const* const name`. Still comes as a reflex to write it this way.

File: include/engine/guidance/processing/segment_compression.hpp [hunk 42]
```diff
@@ -0,0 +1,69 @@
+#ifndef ENGINE_GUIDANCE_PROCESSING_SEGMENT_COMPRESSION_HPP_
```
Comment (TheMarex) [inline include/engine/guidance/processing/segment_compression.hpp]: This does not need an own namespace. `engine::guidance` is deep enough.

File: include/engine/guidance/api_response_generator.hpp [hunk 43]
```diff
@@ -0,0 +1,296 @@
+#ifndef ENGINE_GUIDANCE_API_RESPONSE_GENERATOR_HPP_
```
Comment (TheMarex) [inline include/engine/guidance/api_response_generator.hpp]: This should not be in `guidance` at all.
